### PR TITLE
Update search to include rate limits

### PR DESCRIPTION
**Pull request by carkod**
## Not to be merged
Review PR for https://github.com/canonical/canonicalwebteam.search/pull/33

## QA

- For QA, I set the limit to 3 times per minute, per endpoint.
- Go to https://ubuntu-com-12514.demos.haus/ search something in the search bar
- Once you are on that page, refresh a couple times or change the search term and you'll get the rate limit (set to 3/minute), after which you'll have to wait one minute to be able to search again.

Only the /search has `3/minute` limit applied, so for other endpoints you won't see th 429 until you reach `500/day`

## Issue / Card

Fixes #

## Screenshots

[If relevant, please include a screenshot.]

## Help

[QA steps](https://discourse.canonical.com/t/qa-steps/152) - [Commit guidelines](https://discourse.canonical.com/t/commit-guidelines/148)

### CODE REVIEW

File: webapp/app.py [hunk 1]
```diff
@@ -523,15 +523,36 @@ def utility_processor():
...
+from flask_limiter.util import get_remote_address
+
+
+limiter = Limiter(
+    key_func=get_remote_address,
+    app=flask.g
+)
+
+class TestView(flask.views.View):
+    decorators = [limiter.limit("2/minute")]
```
Comment (carkod) [inline webapp/app.py:538]: Need to decide what limit

Comment (nottrobin) [reply webapp/app.py]: Yeah this is a good initial setting for testing, but how should we decide a good limit?

I spent a little while looking into Graylog for the IPs you asked IS to block:

For 78.128.113.62, we got between 25 and 238 per minute, for 11 minutes between 09:05 and 09:26 on January 31st:

image

For 5.196.153.58 it was more sustained over 24 hours between 08:57 on 31st January to 07:27 on 1st February, peaking at 1300/minute, lowest was 3/minute, but probably averaging out around 50/minute:

image

For 103.212.98.234, it was more than 24 hours between 08:10 on 5th February to 10:40 on 6th February. The peak was 949/minute, the low was about 2/minute, and it had this spiking quality, with a spike of around 200/minute every 10 minutes or so.

image

I wonder if it's possible to set multiple thresholds? Maybe we should have a threshold of around 50/minute, but maybe maximum per 10 minutes and per hour?

**Comment by nottrobin** [comment]
> Should we handle 429 pages in the project or module?

The module should do `flask.abort(429)`, and the project should create a 429 template. Might be worth adding a 429 error handing route option [to flask-base](https://github.com/canonical/canonicalwebteam.flask-base/blob/main/canonicalwebteam/flask_base/app.py#L246).

**Comment by nottrobin** [comment]
> Not entirely sure about this, but could squid proxy affect the reliability of flask.request.remote_addr? Which is what is used by flask-limiter to get the IP and apply the limits. If it does we may [need this](https://flask-limiter.readthedocs.io/en/stable/recipes.html#deploying-an-application-behind-a-proxy)

Flask-base [already implements ProxyFix](https://github.com/canonical/canonicalwebteam.flask-base/blob/20e5d5b0f3cfe8c8cd0ea10385a5a9782ddf919f/canonicalwebteam/flask_base/app.py#L211), I believe.

File: templates/400.html [hunk 2]
```diff
@@ -1,4 +1,4 @@
-{% extends "templates/one-column.html" %} {% block title %}404: Page not found{%
+{% extends "templates/one-column.html" %} {% block title %}400: Bad Request{%
```
Comment (nottrobin) [inline templates/400.html:1]: :raised_hands:

File: webapp/app.py [hunk 3]
```diff
@@ -219,6 +219,15 @@ def deleted_error(error):
     return flask.render_template("410.html", message=error.description), 410
 
 
+@app.errorhandler(429)
+def too_many_requests(error):
+    """
+    Endpoint abuse error
+    """
+    custom_error = f'{error.description}. Please try again tomorrow.'
```
Comment (nottrobin) [inline webapp/app.py:227]: In QAing I realised this message won't work any more. I got "The rate limit is: 3/minute. Please try again tomorrow.". It'll have to be "Please try again later."
```suggestion
    custom_error = f'{error.description}. Please try again later.'
```

**Review by nottrobin — APPROVED** [review]
LGTM :+1:. Just need to remove the rate limit override, and change the requirements.txt to use the new version, once it's published. Oh, and fix the linter.

**Pull request merged**